src/components/pages/users.js

```diff
@@ -121,116 +167,106 @@ class Users extends React.Component {
     return <b>{username}</b>;
   }
 
-  renderRoleTags(userRoles, user) {
-    return availableRoles.map(ar => {
+  renderRoleTags(_userRoles, user) {
+    return availableRoles.map(role => {
       return (
         <UserRoleTagEditor
-          key={ar.name}
+          key={role}
           user={user}
-          roleName={ar.name}
-          isReadonly={ar.isReadonly}
+          roleName={role}
           onRoleChange={this.handleRoleChange}
           />
       );
     });
   }
 
-  renderLockedOutState(lockedOut, user) {
+  renderLockedOutState(_lockedOut, user) {
     return <UserLockedOutStateEditor user={user} onLockedOutStateChange={this.handleLockedOutStateChange} />;
   }
 
-  handleRoleChange(user, newRoles) {
-    this.setState(prevState => {
-      return {
-        ...prevState,
-        users: prevState.users.map(usr => usr._id === user._id ? { ...user, roles: newRoles } : usr),
-        changedRoles: {
-          ...prevState.changedRoles,
-          [user._id]: newRoles
-        },
-        isDirty: true
-      };
-    });
-  }
-
-  handleLockedOutStateChange(user, newLockedOut) {
-    this.setState(prevState => {
-      return {
-        ...prevState,
-        users: prevState.users.map(usr => usr._id === user._id ? { ...user, lockedOut: newLockedOut } : usr),
-        changedLockOutStates: {
-          ...prevState.changedLockOutStates,
-          [user._id]: newLockedOut
-        },
-        isDirty: true
-      };
-    });
+  handleTabClick(evt) {
+    this.setState({ currentTab: evt.key });
   }
 
-  async handleSaveClick() {
+  async handleRoleChange(user, newRoles) {
     const { userApiClient, t } = this.props;
-    const { changedLockOutStates, changedRoles } = this.state;
+    const oldRoles = user.roles;
 
-    try {
-      for (const [userId, newLockedOut] of Object.entries(changedLockOutStates)) {
-        /* eslint-disable-next-line no-await-in-loop */
-        await userApiClient.saveUserLockedOutState({ userId, lockedOut: newLockedOut });
-      }
+    this.setState(prevState => ({
+      ...prevState,
+      internalUsers: prevState.internalUsers.map(usr => usr._id === user._id ? { ...user, roles: newRoles } : usr),
+      isSaving: true
```

Comment:
Extracting 
`prevState.internalUsers.map(usr => usr._id === user._id ? { ...user, roles: newRoles } : usr)`
in a const would make this a bit more readable I find. Maybe also less confusing between `usr` and `user` to call `usr` -> `internalUser` then the mapping gets simpler to understand from one look.

Reply:
I tried several variations and ended up with a helper function `replaceUserById(users, newUser)` that makes the line (and the other three occurrences) less cryptic.

Reply:
Sounds good!